namer/namer.cc

```diff
@@ -1175,18 +1179,28 @@ class SymbolDefiner {
             symbol.data(ctx)->setSuperClass(core::Symbols::Net_Protocol());
         }
 
-        // Skip adding locs when kind is Unknown, becaue that means it was a only a FoundClass for a
-        // class or static field scope (not the class def itself), and we want to treat those as
-        // usage locs, not definition locs.
-        // TODO(jez) This causes known problems on the fast path, where these locs can fail to be
-        // updated and crash. We've chosen the current approach because (1) it matches old behavior
-        // (2) adding O(files) locs for something like Opus is far too slow.
         const bool isUnknown = klass.classKind == core::FoundClass::Kind::Unknown;
+
         // Don't add locs for <root>; 1) they aren't useful and 2) they'll end up with O(files in
         // project) locs!
-        if (symbol != core::Symbols::root() && !isUnknown) {
-            symbol.data(ctx)->addLoc(ctx, ctx.locAt(klass.declLoc));
-            if (klass.definesBehavior) {
+        if (symbol != core::Symbols::root()) {
+            // If the kind is unknown, it means it was only a FoundClass for a class or static field scope, not
+            // the class def itself. We want to generally treat these as usage locs, not definition locs.
+            //
+            // At best, we only want to keep one loc in the codebase per unknown class for perf reasons. We don't
+            // want to store O(files) locs for something like Opus. So if the existing loc (which would have been
+            // brought into existence by getClassSymbol()) is not from this file, we don't add a new loc.
+            //
+            // If the unknown class loc is from the same file, it's still possible that it is from a real
+            // definition in that file. In which case, we check the declared bit on the class.
+            // We only set the loc if the class is not declared.
+            bool updateLoc = !isUnknown || (isUnknown && symbol.data(ctx)->loc().file() == ctx.file &&
+                                            !symbol.data(ctx)->isDeclared());
```

Comment:
```suggestion
            bool updateLoc = !isUnknown || (!symbol.data(ctx)->isDeclared() &&
                                            symbol.data(ctx)->loc().file() == ctx.file);
```

this reads slightly better IMO